### DISCUSSION TITLE
Follow Camera for 3D in FGXL 17

### DISCUSSION THREAD

**Discussion by taranion** [AlmasB/FXGL#1235] [Q&A]
Hey,
I am just learning my way around in FXGL and I am trying a 3D game. Basically I want the camera to follow the 3D view of the player, being in a fixed postion in relation to the player entity, while the player entity is moving around.

In pure JavaFX I would put a perspective camera and the player node in a group and move the group. But it seems FXGL has a dedicated camera handling and I don't know how to do it in FXGL.

Can anybody help?

**Comment by AlmasB** [comment]
Hi,

you can get the 3D camera instance from `FXGL.getGameScene().*`. Then, `camera.getTransformComponent()` gives you the camera transform, which can be manipulated like any other entity, so it should provide sufficient functionality to place, rotate and keep the camera as required.

Give a shout if you get stuck.

**Reply by taranion** [reply]
Hey Almas,
thank you, that sort of worked. I bound the properties of the camera to those of the player.
```
camera3D.getTransform().xProperty().bind(player.xProperty());
camera3D.getTransform().yProperty().bind(player.yProperty().subtract(150));
camera3D.getTransform().zProperty().bind(player.zProperty().subtract(400));
camera3D.getTransform().lookAt(player.getPosition3D());
```
I can follow the player now - as long as I am just going forward.

But when I turn in any direction and rotate the player,  I need the camera to automatically move in a position that is still behind the player, even though it is rotated. Of course it should be possible to calculate that position, but that would defeat the point of binding the properties to the player. 
If I understand the SceneGraph currectly, grouping player and camera into a javafx.scene.Group and performing transformation actions on the group should do the trick. Is there a way to do that with FXGL? Or do you have any other idea that would spare me the 3D vector math?

**Reply by AlmasB** [reply]
There is no built-in way of doing it yet. Essentially, as I understand you'd like to add a third-person camera that follows the player? Could you point me to an example game perhaps, as there are multiple ways a third-person camera can behave. It would be good to have this functionality built into FXGL, similar to `.setFPSCamera()`.

In the meantime, perhaps the following achieves what you are trying to do?

```
    @Override
    protected void onUpdate(double tpf) {
        var distanceToCamera = 8;
        var pos = player.getPosition3D().subtract(player.getTransformComponent().getDirection3D().multiply(distanceToCamera));

        transform.setPosition3D(pos);
        transform.lookAt(player.getPosition3D());
    }
```

**Reply by AlmasB** [reply]
I should note that this is using the following dev build. I think the latest stable may have a bug related to `getDirection3D()`.

To get the latest build, you can use this:
```
<dependencies>
    <dependency>
        <groupId>com.github.almasb</groupId>
        <artifactId>fxgl</artifactId>
        <version>17+dev-SNAPSHOT</version>
    </dependency>
</dependencies>

<repositories>
    <repository>
        <id>oss.sonatype.org-snapshot</id>
        <url>https://oss.sonatype.org/content/repositories/snapshots</url>
    </repository>
</repositories>
```

**Reply by taranion** [reply]
Thanks for your response. I was already doing experimental stuff like moving the camera in the scenegraph :)
```
PerspectiveCamery camera = getGameScene().getCamera3D().getPerspectiveCamera();
player.getViewComponent().addChild(camera, true);
```
but that required unexpected translations to work. (Though it did somehow).

> Essentially, as I understand you'd like to add a third-person camera that follows the player? 

That is correct. My application is more or less a flight simulator in space and want a camera that has a fixed position in relation to the player - from a slightly elevated position looking down at the player.
The player ship needs to be able to rotate on all 3 axis (see here: https://builtbyaz.com/simpit/flight-sim-rudder-pedals/ )

Your example works, but I think there is still a problem with Direction3D. As long as I only rotate around the X- or Y-axis (going left/right, up/down), but it doesn't seem to detect rotation around the Z-Axis (transform.setRotationZ). The player rotates correctly, but the cam doesn't follow.

**Comment by Snoopy137** [comment]
Have you figure this one out? I've came up with an implemention that uses only FXGL tools, no extra libraries needed, I'm no expert so it's possible that is not that good, but I find a way to use the FPS camera and update the position when it moves to always look at an entity, of course it can be improved and I could use help with the tweaks so if either of you would like to take a look let me know

**Reply by taranion** [reply]
I'd like to have a look. 
My own solution is working, but so far, but now I hit the problem that rotating on the X axis (Pitch) to either -90 or +90 degrees, flips the model. I noticed that JMonkeyEngine for example uses Quaternions to solve this problem, but being a noob to 3D graphics, I just failed to wrap my head around the problem and reimplement it.

**Reply by Snoopy137** [reply]
Hey @taranion, 
So what I did to get the third person camera is to setFPSCamera(true)and  move the camera to face the entity when it rotates, the code is the following

`double xset = distance * Math.sin(Math.toRadians(camera.getTransform().getRotationY())) * -1 + player.getX();
 double zset = distance * Math.cos(Math.toRadians(camera.getTransform().getRotationY())) * -1 + player.getZ();
 double yset = distance * Math.sin(Math.toRadians(camera.getTransform().getRotationX())) + player.getY() + player.getScaleY() - 2;
 camera.getTransform().setPosition3D(xset, yset, zset);`

the problem I have with this is the object direction, I use the animation builder to rotate the object X and Y axis and make it look in the same direction as the camera but fore some reason this does not update the onject direction3d, don't know if this is a bug or not, a workaround for moving forward the object is tu use the following code

`player.getTransformComponent().translate3D(camera.getTransform().getDirection3D());`

but I still haven't found a way to move to the sides or backwards

**Reply by Snoopy137** [reply]
Hey @taranion  I made some really good progress on this one
So to avoid the player to flip when rotating is best to use animationbuilder, I use the following method
`private void playerRotate() {
        animationBuilder()
                .duration(Duration.millis(150))
                .autoReverse(false)
                .rotate(player)
                .from(new Point3D(player.getTransformComponent().getRotationX(), player.getTransformComponent().getRotationY(), player.getTransformComponent().getRotationZ()))
                .to(new Point3D(camera.getTransform().getRotationX(), camera.getTransform().getRotationY(), camera.getTransform().getRotationZ()))
                .buildAndPlay();
    }`

with this the entity rotates on X and Y axis to face the same direction as the camera, still if you call the entity.gettransfor.getdirection3d it will say the entity directino is x=0 y=0 z=0.
To solve this and have the entity set in the same direction3d as the camera I use this two lines:
`Point3D look = camera.getTransform().getDirection3D().add(player.getPosition3D());
        player.getTransformComponent().lookAt(look);`
with thtat you can use moveforward, moveleft and so on to move the player, camera will always follow along if you have added the lines from my previous comment on the onupdate method.
Missing details are, you can only move the player relative to the camera direction, so if you want to have the player move forward and have the camera looking on player side is not possible with this approach, I'm still working on that one, but I think this is very useful though.
Please take look at the full code and let me know what you think.
[Third person camera test](https://github.com/Snoopy137/FXGLthirdpersoncamera.git)

Regards

PS: @AlmasB this code could be significantlly simplfy if the entity direction3d is updated when is rotated by the animation builder, it would also be possible to have a third person camera that can look at the player indpenedantly from it's direction, similar to the one from Resident Evil 4, don't know if this is a bug with entities direction3d or I'm just not understanding its workflow but still I think that with a very little tweak this could be used to have a third person camera out of the box on FXGL
Regards!

**Reply by AlmasB** [reply]
Thanks, will check this out in the upcoming days. The `direction3D` bug should be fixed in the `dev` version, which will eventually be released as 17.3, but to get the latest build now, you can use this:
```
<dependencies>
    <dependency>
        <groupId>com.github.almasb</groupId>
        <artifactId>fxgl</artifactId>
        <version>17+dev-SNAPSHOT</version>
    </dependency>
</dependencies>

<repositories>
    <repository>
        <id>oss.sonatype.org-snapshot</id>
        <url>https://oss.sonatype.org/content/repositories/snapshots</url>
    </repository>
</repositories>
```

**Reply by taranion** [reply]
@Snoopy137 I tested your third person camera test. It helps understand why you are using the AnimationBuilder  ... having a slight delay between camera and the ship is nice - though I would have expected it the other way around: let the ship lead and the camera follow.

And another thing that I am missing: If you look up or down 90°C, you don't see the ship anymore. The camera looks down/up from the position where it was following. This is great for "looking around with the camera", but I want to enforce a view from behind the ship.
So I need the camera to go up and forward.

But ignoring my needs: You demo works well. It makes me reconsider to use the FPSCamera (I decided against it before).

**Reply by Snoopy137** [reply]
Hi @taranion , yes, I noticed the look up and down, the math there needs to be tweaked, I just need to figure the right count for that.

If understand correctly you need that the camera is always behind the ship, well, if that's the case, you can simply rotate the camera at the same time and angle as the ship, the math will put the camera behind it

Let me know if that works or an example or how you want the camera

**Reply by Snoopy137** [reply]
@taranion the 90 degrees rotation problem has been fixed, you can check the code
Regards!

**Comment by taranion** [comment]
@AlmasB By the way, you asked
> Could you point me to an example game perhaps, as there are multiple ways a third-person camera can behave. It would be good to have this functionality built into FXGL, similar to .setFPSCamera().

I figured that this matches my needs very closely:
```
getGameScene().setFPSCamera(false);
player.getViewComponent().addChild(camera.getPerspectiveCamera(), true);
```
I needed to move the camera a bit closer to my ship
```
camera.getPerspectiveCamera().setTranslateZ(7);
```
That creates a view, like the camera being on a fixed stick behind the ship, copying all rotations and being in a fixed distance - which is what I need. 
Now the only thing that would be nice, if I find a way to temporarily rotate the model in the player view component in the direction the player just chose.

**Comment by taranion** [comment]
@Snoopy137 I took the liberty to fork your repo and modify it to show what I intend to do
https://github.com/taranion/FXGLSpaceshipFlightTest

Control with
- W,S : Forward/Backward
- A,D : Turn Left/Right (Yaw)
- Up, Down: Turn Up, Down (Pitch)
- Left, Right : Roll left, right
